dnf-behave-tests/features/downgrade.feature

```diff
@@ -41,3 +41,82 @@ Scenario: Downgrade RPM that requires downgrade of dependency
         | downgrade     | glibc-0:2.28-9.fc29.x86_64                |
         | downgrade     | glibc-common-0:2.28-9.fc29.x86_64         |
         | downgrade     | glibc-all-langpacks-0:2.28-9.fc29.x86_64  |
+
+
+@bz1759847
+Scenario: Check downgrade / upgrade exit codes
+  Given I use repository "dnf-ci-fedora"
+    And I use repository "dnf-ci-fedora-updates"
+   # attempt to downgrade/upgrade non existent package exits 1
+   When I execute dnf with args "downgrade non-existent-package"
+   Then the exit code is 1
+    And stdout is
+    """
+    <REPOSYNC>
+    No package non-existent-package available.
+    """
+    And stderr is
+    """
+    Error: No packages marked for downgrade.
+    """
+   When I execute dnf with args "upgrade non-existent-package"
+   Then the exit code is 1
+    And stdout is
+    """
+    <REPOSYNC>
+    No match for argument: non-existent-package
+    """
+    And stderr is
+    """
+    Error: No packages marked for upgrade.
+    """
+   # attempt to downgrade/upgrade non installed package exits 1
+   When I execute dnf with args "downgrade flac"
+   Then the exit code is 1
+    And stdout is
+    """
+    <REPOSYNC>
+    Packages for argument flac available, but not installed.
+    """
+    And stderr is
+    """
+    Error: No packages marked for downgrade.
+    """
+   When I execute dnf with args "upgrade flac"
+   Then the exit code is 1
+    And stdout is
+    """
+    <REPOSYNC>
+    No match for argument: flac
+    """
+    And stderr is
+    """
+    Package flac available, but not installed.
+    Error: No packages marked for upgrade.
+    """
+   # attempt to upgrade package already on the highest version
+  Given I successfully execute dnf with args "install flac-0:1.3.3-3.fc29.x86_64"
+   When I execute dnf with args "upgrade flac"
+   Then the exit code is 0
+    And stdout is
+    """
+    <REPOSYNC>
+    Dependencies resolved.
+    Nothing to do.
+    Complete!
+    """
+   # attempt to downgrade package already on the lowest version
+  Given I successfully execute dnf with args "install flac-0:1.3.2-8.fc29.x86_64"
+   When I execute dnf with args "downgrade flac"
+   Then the exit code is 0
+    And stdout is
+    """
+    <REPOSYNC>
+    Dependencies resolved.
+    Nothing to do.
+    Complete!
+    """
+    And stderr is
+    """
+    Package flac of lowest version already installed, cannot downgrade it.
+    """
```

Comment:
Wouldn't this be better split into multiple scenarios? The main difference being the scenarios would be run independently, here the first fail causes the rest not to be run. Since this even tests two different commands (upgrade/downgrade), I think it would be better to split into smaller pieces.

For the common setup these tests have, `Given I use repository "dnf-ci-fedora-updates"` is shared with all the scenarios, so that can go into `Background`. And then `"dnf-ci-fedora"` is only required for a couple of tests it seems.

Otherwise LGTM.

Reply:
One of the goals for this test was to make sure that upgrade/downgrade commands behave consistently - that they return the same exit code in the same situations. So I would like to keep them together. But I agree the scenario could be split into three (four) independent tests - non existent package, not installed package, already on highest (lowest) version. The background bit of course could be done.

Reply:
Eventually I've split them into multiple scenarios.